cpp/src/arrow/engine/substrait/serde.cc

```diff
@@ -58,12 +58,57 @@ Result<compute::Declaration> DeserializeRelation(const Buffer& buf,
   return FromProto(rel, ext_set);
 }
 
-Result<std::vector<compute::Declaration>> DeserializePlans(
-    const Buffer& buf, const ConsumerFactory& consumer_factory,
+using DeclarationFactory = std::function<compute::Declaration(
+    compute::Declaration, std::vector<std::string> names)>;
+
+static DeclarationFactory MakeConsumingSinkDeclarationFactory(
+    const ConsumerFactory& consumer_factory) {
+  return [&consumer_factory](compute::Declaration input, std::vector<std::string> names) {
+    std::shared_ptr<compute::ExecNodeOptions> options =
+        std::make_shared<compute::ConsumingSinkNodeOptions>(
+            compute::ConsumingSinkNodeOptions{consumer_factory(), std::move(names)});
+    return compute::Declaration::Sequence(
+        {std::move(input), {"consuming_sink", options}});
+  };
+}
+
+namespace {
+
+compute::Declaration ProjectByNamesDeclaration(compute::Declaration input,
+                                               std::vector<std::string> names) {
+  int names_size = static_cast<int>(names.size());
+  if (names_size == 0) {
+    return input;
+  }
+  std::vector<compute::Expression> expressions;
+  for (int i = 0; i < names_size; i++) {
+    expressions.push_back(compute::field_ref(FieldRef(i)));
+  }
+  return compute::Declaration::Sequence(
+      {std::move(input),
+       {"project",
+        compute::ProjectNodeOptions{std::move(expressions), std::move(names)}}});
+}
+
+}  // namespace
+
+static DeclarationFactory MakeWriteDeclarationFactory(
+    const WriteOptionsFactory& write_options_factory) {
+  return [&write_options_factory](compute::Declaration input,
+                                  std::vector<std::string> names) {
+    compute::Declaration projected = ProjectByNamesDeclaration(input, names);
+    std::shared_ptr<compute::ExecNodeOptions> options = write_options_factory();
+    return compute::Declaration::Sequence({std::move(projected), {"write", options}});
```

Comment:
In a recent PR, I added `names` to `ConsumingSinkNodeOptions` for a similar reason. Given your comment here, I suspect it may be useful to refactor this names-feature (perhaps into its own options instance to be included in more complex options instances). If you agree, then the JIRA should be for this feature.

Reply:
I agree with you.  Every "sink" node (including writes) should probably allow the user to provide a schema for the outgoing batches as we convert from ExecBatch to RecordBatch at the edge.  This would allow the user to supply custom names as well as custom field and schema-level metadata.

The only thing they can't specify is the field types (those have to match what is already the output schema for the node).  So we can either:

 * Use schema and throw an error if the user-supplied schema's types don't match
 * Create a new type that is just the names & metadata and not types.

Reply:
Please create a JIRA for this. Perhaps there should be a hierarchy - names only, names plus metadata, and schema - in increasing order of detail. The names-only option is what's needed for the Substrait plan. If the caller provides a schema then the types must match, as you say.

Reply:
ARROW-16915